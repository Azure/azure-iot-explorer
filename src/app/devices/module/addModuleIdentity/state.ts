/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';
import { AddDeviceStateInterface } from '../../addDevice/state';
import { IM } from '../../../shared/types/types';

export interface AddModuleStateInterface {
    synchronizationStatus: SynchronizationStatus;
}

export const addModuleStateInitial = Record<AddDeviceStateInterface>({
    synchronizationStatus: SynchronizationStatus.initialized
});

export type AddModuleStateType = IM<AddModuleStateInterface>;
