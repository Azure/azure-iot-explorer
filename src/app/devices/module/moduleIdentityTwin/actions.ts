/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { DEVICECONTENT } from '../../../constants/actionPrefixes';
import { GET_MODULE_IDENTITY_TWIN, UPDATE_MODULE_IDENTITY_TWIN } from '../../../constants/actionTypes';
import { ModuleTwin } from '../../../api/models/moduleTwin';

export const getModuleIdentityTwinAction = actionCreatorFactory(DEVICECONTENT).async<GetModuleIdentityTwinActionParameters, ModuleTwin>(GET_MODULE_IDENTITY_TWIN);
export const updateModuleIdentityTwinAction = actionCreatorFactory(DEVICECONTENT).async<ModuleTwin, ModuleTwin>(UPDATE_MODULE_IDENTITY_TWIN);

export interface GetModuleIdentityTwinActionParameters {
    moduleId: string;
    deviceId: string;
}
