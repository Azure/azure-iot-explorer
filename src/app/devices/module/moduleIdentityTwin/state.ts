/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../../shared/types/types';
import { SynchronizationWrapper } from '../../../api/models/synchronizationWrapper';
import { ModuleTwin } from '../../../api/models/moduleTwin';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';

export interface ModuleTwinStateInterface extends SynchronizationWrapper<ModuleTwin> {}

export const moduleTwinStateInitial = Record<ModuleTwinStateInterface>({
    payload: undefined,
    synchronizationStatus: SynchronizationStatus.initialized
});

export type ModuleTwinStateType = IM<ModuleTwinStateInterface>;
