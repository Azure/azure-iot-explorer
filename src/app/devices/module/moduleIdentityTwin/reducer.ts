/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { reducerWithInitialState } from 'typescript-fsa-reducers';
import { moduleTwinStateInitial, ModuleTwinStateType } from './state';
import {
    getModuleIdentityTwinAction,
    updateModuleIdentityTwinAction,
    GetModuleIdentityTwinActionParameters,
} from './actions';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';
import { ModuleTwin } from '../../../api/models/moduleTwin';

export const moduleTwinReducer = reducerWithInitialState<ModuleTwinStateType>(moduleTwinStateInitial())
    .case(getModuleIdentityTwinAction.started, (state: ModuleTwinStateType) => {
        return state.merge({
            synchronizationStatus: SynchronizationStatus.working
        });
    })
    .case(getModuleIdentityTwinAction.done, (state: ModuleTwinStateType, payload: {params: GetModuleIdentityTwinActionParameters} & {result: ModuleTwin}) => {
        return state.merge({
            payload: payload.result,
            synchronizationStatus: SynchronizationStatus.fetched
        });
    })
    .case(getModuleIdentityTwinAction.failed, (state: ModuleTwinStateType) => {
        return state.merge({
            synchronizationStatus: SynchronizationStatus.failed
        });
    })
    .case(updateModuleIdentityTwinAction.started, (state: ModuleTwinStateType) => {
        return state.merge({
            synchronizationStatus: SynchronizationStatus.updating
        });
    })
    .case(updateModuleIdentityTwinAction.done, (state: ModuleTwinStateType, payload: {params: ModuleTwin} & {result: ModuleTwin}) => {
        return state.merge({
            payload: payload.result,
            synchronizationStatus: SynchronizationStatus.upserted
        });
    })
    .case(updateModuleIdentityTwinAction.failed, (state: ModuleTwinStateType) => {
        return state.merge({
            synchronizationStatus: SynchronizationStatus.failed
        });
    });
