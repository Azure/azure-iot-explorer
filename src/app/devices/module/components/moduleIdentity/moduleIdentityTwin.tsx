/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { RouteComponentProps, Route } from 'react-router-dom';
import { CommandBar } from 'office-ui-fabric-react/lib/CommandBar';
import { SpinnerSize, Spinner } from 'office-ui-fabric-react/lib/Spinner';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../shared/contexts/localizationContext';
import { ThemeContextConsumer, ThemeContextInterface } from '../../../../shared/contexts/themeContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { REFRESH, NAVIGATE_BACK } from '../../../../constants/iconNames';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { ROUTE_PARTS, ROUTE_PARAMS } from '../../../../constants/routes';
import { GetModuleIdentityTwinActionParameters } from '../../actions';
import { ModuleTwin } from '../../../../api/models/moduleTwin';
import MultiLineShimmer from '../../../../shared/components/multiLineShimmer';
import { ModuleIdentityDetailHeaderContainer } from './moduleIdentityDetailHeaderView';
import '../../../../css/_deviceDetail.scss';
import '../../../../css/_moduleIdentityDetail.scss';

const EditorPromise = import('react-monaco-editor');
const Editor = React.lazy(() => EditorPromise);

export interface ModuleIdentityTwinDataProps {
    moduleIdentityTwin: ModuleTwin;
    moduleIdentityTwinSyncStatus: SynchronizationStatus;
}

export interface ModuleIdentityTwinDispatchProps {
    getModuleIdentityTwin: (params: GetModuleIdentityTwinActionParameters) => void;
}

export type ModuleIdentityDetailProps = ModuleIdentityTwinDataProps & ModuleIdentityTwinDispatchProps;
export default class ModuleIdentityDetailComponent
    extends React.Component<ModuleIdentityDetailProps & RouteComponentProps> {

    public render(): JSX.Element {
        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <>
                        {this.showCommandBar(context)}
                        <Route component={ModuleIdentityDetailHeaderContainer} />
                        <div className="module-identity-detail">
                            {this.props.moduleIdentityTwinSyncStatus === SynchronizationStatus.working ?
                                <MultiLineShimmer/> :
                                this.showModuleTwin()
                            }
                        </div>
                    </>
                )}
            </LocalizationContextConsumer>
        );
    }

    public componentDidMount() {
        this.retrieveData();
    }

    public componentDidUpdate(oldProps: ModuleIdentityTwinDataProps & RouteComponentProps) {
        if (getModuleIdentityIdFromQueryString(oldProps) !== getModuleIdentityIdFromQueryString(this.props)) {
            this.retrieveData();
        }
    }

    private readonly retrieveData = () => {
        const deviceId = getDeviceIdFromQueryString(this.props);
        const moduleId = getModuleIdentityIdFromQueryString(this.props);
        this.props.getModuleIdentityTwin({
            deviceId,
            moduleId
        });
    }

    private readonly showCommandBar = (context: LocalizationContextInterface) => {
        return (
            <CommandBar
                className="command"
                items={[
                    {
                        ariaLabel: context.t(ResourceKeys.moduleIdentity.detail.command.refresh),
                        disabled: this.props.moduleIdentityTwinSyncStatus === SynchronizationStatus.working,
                        iconProps: {iconName: REFRESH},
                        key: REFRESH,
                        name: context.t(ResourceKeys.moduleIdentity.detail.command.refresh),
                        onClick: this.retrieveData
                    }
                ]}
                farItems={[
                    {
                        ariaLabel: context.t(ResourceKeys.moduleIdentity.detail.command.back),
                        iconProps: {iconName: NAVIGATE_BACK},
                        key: NAVIGATE_BACK,
                        name: context.t(ResourceKeys.moduleIdentity.detail.command.back),
                        onClick: this.navigateToModuleList
                    }
                ]}
            />
        );
    }

    private readonly showModuleTwin = () => {
        return (
            <>
                {this.props.moduleIdentityTwin &&
                    <div className="monaco-editor">
                        <React.Suspense fallback={<Spinner title={'loading'} size={SpinnerSize.large} />}>
                            <ThemeContextConsumer>
                                {(themeContext: ThemeContextInterface) => (
                                    <Editor
                                        language="json"
                                        value={JSON.stringify(this.props.moduleIdentityTwin, null, '\t')}
                                        options={{
                                            automaticLayout: true,
                                            readOnly: true
                                        }}
                                        theme={themeContext.monacoTheme}
                                    />
                                )}
                            </ThemeContextConsumer>
                        </React.Suspense>
                    </div>
                }
            </>
        );
    }

    private readonly navigateToModuleList = () => {
        const path = this.props.match.url.replace(/\/moduleIdentity\/moduleTwin\/.*/, `/${ROUTE_PARTS.MODULE_IDENTITY}`);
        const deviceId = getDeviceIdFromQueryString(this.props);
        this.props.history.push(`${path}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`);
    }
}
