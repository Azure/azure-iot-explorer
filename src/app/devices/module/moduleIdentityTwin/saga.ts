/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { all, takeEvery, takeLatest } from 'redux-saga/effects';
import { getModuleIdentityTwinAction, updateModuleIdentityTwinAction } from './actions';
import { getModuleIdentityTwinSaga } from './sagas/getModuleIdentityTwinSaga';
import { updateModuleIdentityTwinSaga } from './sagas/updateModuleIdentityTwinSaga';

export function* moduleIdentityTwinSagas() {
    yield all([
        takeLatest(getModuleIdentityTwinAction.started.type, getModuleIdentityTwinSaga),
        takeEvery(updateModuleIdentityTwinAction.started.type, updateModuleIdentityTwinSaga)
    ]);
}
