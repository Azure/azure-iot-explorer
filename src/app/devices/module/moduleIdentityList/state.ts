/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../../shared/types/types';
import { SynchronizationWrapper } from '../../../api/models/synchronizationWrapper';
import { ModuleIdentity } from '../../../api/models/moduleIdentity';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';

export interface ModuleIndentityListStateInterface extends SynchronizationWrapper<ModuleIdentity[]>{}

export const moduleIndentityListStateInitial = Record<ModuleIndentityListStateInterface>({
    payload: [],
    synchronizationStatus: SynchronizationStatus.initialized
});

export type ModuleIndentityListStateType = IM<ModuleIndentityListStateInterface>;
