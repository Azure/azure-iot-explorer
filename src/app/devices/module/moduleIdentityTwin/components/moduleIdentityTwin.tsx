/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useHistory } from 'react-router-dom';
import { CommandBar } from '@fluentui/react';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { REFRESH, SAVE } from '../../../../constants/iconNames';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { getModuleIdentityTwinAction, updateModuleIdentityTwinAction } from '../actions';
import { MultiLineShimmer } from '../../../../shared/components/multiLineShimmer';
import { useAsyncSagaReducer } from '../../../../shared/hooks/useAsyncSagaReducer';
import { JSONEditor } from '../../../../shared/components/jsonEditor';
import { moduleTwinReducer } from '../reducer';
import { moduleIdentityTwinSagas } from '../saga';
import { moduleTwinStateInitial } from '../state';
import '../../../../css/_deviceDetail.scss';
import { AppInsightsClient } from '../../../../shared/appTelemetry/appInsightsClient';
import { TELEMETRY_PAGE_NAMES, TELEMETRY_USER_ACTIONS } from '../../../../../app/constants/telemetry';

export const ModuleIdentityTwin: React.FC = () => {
    const { t } = useTranslation();
    const { search, pathname } = useLocation();
    const history = useHistory();
    const moduleId = getModuleIdentityIdFromQueryString(search);
    const deviceId = getDeviceIdFromQueryString(search);

    const [ localState, dispatch ] = useAsyncSagaReducer(moduleTwinReducer, moduleIdentityTwinSagas, moduleTwinStateInitial(), 'moduleIdentityTwinState');
    const moduleIdentityTwin = localState.payload;
    const moduleIdentityTwinSyncStatus = localState.synchronizationStatus;
    const [ state, setState ] = React.useState({
        isDirty: false,
        isTwinValid: true,
        twin: JSON.stringify(moduleIdentityTwin, null, '\t')
    });

    React.useEffect(() => {
        retrieveData();
    },              [deviceId, moduleId]);

    React.useEffect(() => {
        AppInsightsClient.getInstance()?.trackPageView({name: TELEMETRY_PAGE_NAMES.MODULE_TWIN});
    }, []); // tslint:disable-line: align

    const retrieveData = () => dispatch(getModuleIdentityTwinAction.started({ deviceId, moduleId }));

    const handleSave = () => {
        setState({
            ...state,
            isDirty: false,
            isTwinValid: true
        });
        AppInsightsClient.trackUserAction(TELEMETRY_USER_ACTIONS.UPDATE_MODULE_TWIN);
        dispatch(updateModuleIdentityTwinAction.started(JSON.parse(state.twin)));
    };

    const showCommandBar = () => {
        return (
            <CommandBar
                className="command"
                items={[
                    {
                        ariaLabel: t(ResourceKeys.moduleIdentity.detail.command.refresh),
                        disabled: moduleIdentityTwinSyncStatus === SynchronizationStatus.working,
                        iconProps: {iconName: REFRESH},
                        key: REFRESH,
                        name: t(ResourceKeys.moduleIdentity.detail.command.refresh),
                        onClick: retrieveData
                    },
                    {
                        ariaLabel: t(ResourceKeys.moduleIdentity.detail.command.save),
                        disabled: !state.isDirty || !state.isTwinValid,
                        iconProps: {iconName: SAVE},
                        key: SAVE,
                        name: t(ResourceKeys.moduleIdentity.detail.command.save),
                        onClick: handleSave
                    }
                ]}
            />
        );
    };

    const showModuleTwin = () => {
        if (moduleIdentityTwinSyncStatus === SynchronizationStatus.working || moduleIdentityTwinSyncStatus === SynchronizationStatus.updating) {
            return <MultiLineShimmer/>;
        }

        return (
            <article className="device-twin">
                {moduleIdentityTwin &&
                    <JSONEditor
                        content={JSON.stringify(moduleIdentityTwin, null, '\t')}
                        className="json-editor"
                        onChange={onChange}
                    />
                }
            </article>
        );
    };

    const onChange = (data: string) => {
        let isTwinValid = true;
        try {
            JSON.parse(data);
        }
        catch  {
            isTwinValid = false;
        }
        setState({
            ...state,
            isDirty: true,
            isTwinValid,
            twin: data
        });
    };

    return (
        <>
            {showCommandBar()}
            <div className="device-detail">
                {moduleIdentityTwinSyncStatus === SynchronizationStatus.working ?
                    <MultiLineShimmer/> :
                    showModuleTwin()
                }
            </div>
        </>
    );
};
