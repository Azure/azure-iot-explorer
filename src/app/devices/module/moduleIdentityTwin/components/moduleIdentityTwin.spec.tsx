/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import 'jest';
import { shallow } from 'enzyme';
import { ModuleIdentityTwin } from './moduleIdentityTwin';
import * as AsyncSagaReducer from '../../../../shared/hooks/useAsyncSagaReducer';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';
import { ModuleTwin } from '../../../../api/models/moduleTwin';
import { moduleTwinStateInitial } from '../state';

const pathname = 'https://127.0.0.1:3000/#/resources/testhub.azure-devices.net/devices/deviceDetail/moduleIdentity/moduleTwin/?deviceId=newdevice&moduleId=moduleId';
jest.mock('react-router-dom', () => ({
    useHistory: () => ({ push: jest.fn() }),
    useLocation: () => ({ search: '?deviceId=newdevice&moduleId=moduleId', pathname }),
}));

const moduleIdentityTwin: ModuleTwin = {
    authenticationType:'sas',
    cloudToDeviceMessageCount: 0,
    connectionState: 'Disconnected',
    deviceEtag: 'AAAAAAAAAAE=',
    deviceId: 'deviceId',
    etag: 'AAAAAAAAAAE=',
    lastActivityTime: '0001-01-01T00:00:00Z',
    moduleId: 'moduleId',
    properties: {},
    status: 'enabled',
    statusUpdateTime: '0001-01-01T00:00:00Z',
    version: 1,
    x509Thumbprint:  {primaryThumbprint: null, secondaryThumbprint: null},
};

const initialState = {
    ...moduleTwinStateInitial(),
    payload: moduleIdentityTwin
};

describe('moduleIdentityTwin', () => {
    context('snapshot', () => {
        it('matches snapshot while loading', () => {
            jest.spyOn(AsyncSagaReducer, 'useAsyncSagaReducer').mockReturnValue([{ synchronizationStatus: SynchronizationStatus.working }, jest.fn()]);
            expect(shallow(<ModuleIdentityTwin/>)).toMatchSnapshot();
        });

        it('matches snapshot after module twin is fetched', () => {
            jest.spyOn(AsyncSagaReducer, 'useAsyncSagaReducer').mockReturnValue([initialState, jest.fn()]);
            expect(shallow(<ModuleIdentityTwin/>)).toMatchSnapshot();
        });
    });
});
