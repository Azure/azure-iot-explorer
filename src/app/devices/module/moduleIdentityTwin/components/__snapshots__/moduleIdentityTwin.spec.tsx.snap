// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moduleIdentityTwin snapshot matches snapshot after module twin is fetched 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    items={
      Array [
        Object {
          "ariaLabel": "moduleIdentity.detail.command.refresh",
          "disabled": false,
          "iconProps": Object {
            "iconName": "Refresh",
          },
          "key": "Refresh",
          "name": "moduleIdentity.detail.command.refresh",
          "onClick": [Function],
        },
        Object {
          "ariaLabel": "moduleIdentity.detail.command.save",
          "disabled": true,
          "iconProps": Object {
            "iconName": "Save",
          },
          "key": "Save",
          "name": "moduleIdentity.detail.command.save",
          "onClick": [Function],
        },
      ]
    }
  />
  <div
    className="device-detail"
  >
    <article
      className="device-twin"
    >
      <JSONEditor
        className="json-editor"
        content="{
	\\"authenticationType\\": \\"sas\\",
	\\"cloudToDeviceMessageCount\\": 0,
	\\"connectionState\\": \\"Disconnected\\",
	\\"deviceEtag\\": \\"AAAAAAAAAAE=\\",
	\\"deviceId\\": \\"deviceId\\",
	\\"etag\\": \\"AAAAAAAAAAE=\\",
	\\"lastActivityTime\\": \\"0001-01-01T00:00:00Z\\",
	\\"moduleId\\": \\"moduleId\\",
	\\"properties\\": {},
	\\"status\\": \\"enabled\\",
	\\"statusUpdateTime\\": \\"0001-01-01T00:00:00Z\\",
	\\"version\\": 1,
	\\"x509Thumbprint\\": {
		\\"primaryThumbprint\\": null,
		\\"secondaryThumbprint\\": null
	}
}"
        onChange={[Function]}
      />
    </article>
  </div>
</Fragment>
`;

exports[`moduleIdentityTwin snapshot matches snapshot while loading 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    items={
      Array [
        Object {
          "ariaLabel": "moduleIdentity.detail.command.refresh",
          "disabled": true,
          "iconProps": Object {
            "iconName": "Refresh",
          },
          "key": "Refresh",
          "name": "moduleIdentity.detail.command.refresh",
          "onClick": [Function],
        },
        Object {
          "ariaLabel": "moduleIdentity.detail.command.save",
          "disabled": true,
          "iconProps": Object {
            "iconName": "Save",
          },
          "key": "Save",
          "name": "moduleIdentity.detail.command.save",
          "onClick": [Function],
        },
      ]
    }
  />
  <div
    className="device-detail"
  >
    <MultiLineShimmer />
  </div>
</Fragment>
`;
